Formula/loki.rb

```diff
@@ -6,6 +6,11 @@ class Loki < Formula
   license "AGPL-3.0-only"
   head "https://github.com/grafana/loki.git", branch: "main"
 
+  livecheck do
+    url :stable
+    strategy :github_latest
+  end
```

Comment:
A regex is preferred here since the `:github_latest` strategy uses requires using the GitHub API, which is limited on CI. See https://docs.brew.sh/Brew-Livecheck#general-guidelines for details. I suggest this (lightly modified from the linked page):

```suggestion
  livecheck do
    url :stable
    regex(/^v(\d+(?:\.\d+)+)$/i)
  end
```

Run `brew livecheck -vd loki` to check that it's getting the right set of versions.

Reply:
Done. But couldn't this strategy be adapted to just seek for latest release by parsing the page rather than querying the API then?

Reply:
For what it's worth, the `GithubLatest` strategy checks the page for the "latest" release on GitHub, so it doesn't use the API. Up to this point, I've specifically avoided using the API because we used to run into rate limiting issues on CI (years ago I modified any checks that used the API to either check the Git tags or the GitHub website).

Reply:
But if `GithubLatest` doesn't use the API, why isn't it better than manually writing a regex? In which case should one be preferred over the other?

Reply:
Without this `livecheck` block, livecheck uses the `Git` strategy with the `stable` URL. This `livecheck` block also implicitly uses the `Git` strategy, as it uses `url :stable` and the `Git` strategy applies to the `stable` URL. All we needed to do here to avoid tags like `helm-loki-3.2.0` was to add a more restrictive regex (i.e., it's not yet necessary to use the `GithubLatest` strategy here).

We only selectively use the `GithubLatest` strategy because, unless I'm mistaken, checking the GitHub website can also hit a rate limit in extreme circumstances (it's very rare in comparison to the API rate limit but I've seen it happen before). The same isn't true for the `Git` strategy because it simply uses `git ls-remote --tags` internally. Besides that, `GithubLatest` involves multiple requests (i.e., redirecting from the "latest" URL to the specific release page) and uses a little more data, so it's a little heavier/slower and this adds up when you're checking thousands of formulae.

[In the long run, I intend to modify the `GithubLatest` strategy to match the version from the `location` header of the response by default, as this will avoid the redirection and make the check a bit lighter overall. I can't do that until after I've merged the forthcoming `options` feature because we need to be able to optionally check the release page HTML for a small number of checks. The `options` feature is at the top of my to-do list and I'm hoping to create a PR for it this week, now that GSoC is over.]

We use the `Git` strategy by default for applicable URLs (e.g., when the `stable` URL is a tag archive, as is the case here). If the `stable` URL was a release asset instead (e.g., `.../releases/download/example-1.2.3.tar.gz`), we would want to use `GithubLatest` (as described above). If upstream creates releases for more than one major/minor version in the same repository (e.g., `2.x` is latest but `1.x` is still maintained alongside it), we end up having to check the first page of releases [using the `prerelease:false` query] (e.g., `https://github.com/Homebrew/brew/releases?q=prerelease%3Afalse`) and match versions from tag URLs (like `GithubLatest` does). The preference for these approaches generally goes `Git`, `GithubLatest`, `PageMatch` checking the GitHub releases page. We start with the preferred approach and fall back to a different approach as necessary. [I'm considering adding an additional `GithubReleases` strategy to replace the last approach, as it's somewhat common for casks.]

That said, this commentary only applies to this type of situation (i.e., we usually don't want to use the `Git` strategy with a `head` URL when the `stable` URL is an archive from the first-party website, for example). There are times where the upstream GitHub repository isn't carefully maintained (e.g., unstable tags also use a stable tag format, the release situation is weird in a way that causes problems for us, etc.) and we end up having to take a different approach (e.g., checking the first-party download page, etc.) but these are the broad strokes. Some `livecheck` blocks are very straightforward (like this one) and others are a bit subjective but I try to maintain some standards (in homebrew/core, at least), as it makes it easier to pull representative examples from existing `livecheck` blocks (and standardization helps to ease our maintenance burden).

Does this help to clarify things?

Reply:
> We only selectively use the `GithubLatest` strategy because `GithubLatest` involves multiple requests (i.e., redirecting from the "latest" URL to the specific release page) and typically uses more data just to return one version (so it's a little heavier/slower and this adds up when you're checking thousands of formulae). Also, unless I'm mistaken, checking the GitHub website can also hit a rate limit in extreme circumstances (it's very rare in comparison to the API rate limit but I've seen it happen before). The `Git` simply uses `git ls-remote --tags` internally, so it can quickly match versions in all tags with less overhead. [However, there are a handful of extreme cases where a repository contains thousands of tags (which grow at a rapid pace) and we use the `GithubLatest` strategy because it's technically more efficient but I note those situation in a comment before the `livecheck` block.]

What's unclear to me is that this seems to say the opposite of what you said in the first paragraph here: https://github.com/Homebrew/homebrew-core/pull/112183#discussion_r986031908 so that I'm still unsure if it hits the API or not. Other than that, yeah that's more clear.

> That said, this commentary only applies to this type of situation (i.e., we usually don't want to use the `Git` strategy with a `head` URL when the `stable` URL is an archive from the first-party website, for example). There are times where the upstream GitHub repository isn't carefully maintained (e.g., unstable tags also use a stable tag format, the release situation is weird in a way that causes problems for us, etc.) and we end up having to take a different approach (e.g., checking the first-party download page, etc.) but these are the broad strokes. Some `livecheck` blocks are very straightforward (like this one) and others are a bit subjective but I try to maintain some standards (in homebrew/core, at least), as it makes it easier to pull representative examples from existing `livecheck` blocks (and standardization helps to ease our maintenance burden).

Yeah some things can make things annoying to maintain. From what I read in some PR here and there resources can now be livechecked which should make it quite easier, though it seems it needs to be manually specified and isn't inferred like for the formula, but I've no idea about why.
I also suggested in an issue some time ago to be able to parametrize urls so that people who bump formula doesn't need to do manual edits, especially this feature already exist for casks, but it seems it wasn't something wishable for everyone.